---
title: "2_blup_prediction"
author: "Hayato Yoshioka"
date: "2025-03-03"
output: html_document
---

# RF for phenotyoe 
from 32 and 63


```{r}
library(MASS)
library(RAINBOWR)
library(foreach)
library(ggplot2)
require(reshape2)
```


```{r}
cd = "Control"

if (cd == "Drought"){
	CD<-"W4"
	Data<-readRDS(here("data/SoyData_Drought2.RDS"))
} else if (cd=="Control"){
	CD<-"W1"
	Data<-readRDS(here("data/SoyData_Control2.RDS"))
}

genome<-Data$genome
met.cd<-Data$met
mic.cd<-Data$mic
mir.cd<-Data$mir
psc19<-Data$pheno
miname<-Data$microname
metname<-Data$metname
cd<-Data$cd

genome<- scale(genome)
met.cd<- scale(met.cd)
mic.cd <- scale(mic.cd)
psc19 <- scale(psc19)

```


```{r, message =F, include=TRUE,eval=TRUE}
K1 <- tcrossprod(genome) / ncol(genome)
K2 <- tcrossprod(mic.cd) / ncol(mic.cd)
K3 <- tcrossprod(met.cd) / ncol(met.cd)

Z <- design.Z(pheno.labels = rownames(genome),geno.names = rownames(genome))

KZ1_2_3<- list(A= list(K = K1, Z = Z),B= list(K = K2, Z = Z),C= list(K = K3, Z = Z))


```


```{R, message =F, include=TRUE,eval=TRUE}
y <- psc19

resblup<- foreach(zl = 1) %do% {
  print("fine")
  ZETA <- KZ1_2_3
  X0 <- NULL
  
  result_matrix <- matrix(nrow = 3*nrow(y), ncol = 0)
  for (i in 1:ncol(y)){
    resEM3 <- EM3.cpp(y = y[,i],
                        X0 = X0, n.core = 1, 
                        ZETA = ZETA)
    result_matrix <- cbind(result_matrix, resEM3[["u.each"]])
  }
  return(result_matrix)
  }
  
names(resblup)<- "KZ1_2_3"
```


```{r}
bget <- function(Xp,K,Up) {
  XXt <- K * ncol(Xp)
  B <- t(Xp) %*% ginv(XXt) %*% Up
  return(B)
}

Up<- resblup[["KZ1_2_3"]]

Up1<-Up[1:nrow(y),]
Up2<-Up[(nrow(y)+1):(nrow(y)*2),]
Up3<-Up[(nrow(y)*2+1): (nrow(y)*3),]

#Xp1<-genome.cd
Xp1<-genome
Xp2<-mic.cd
Xp3<-met.cd


B1<-bget(Xp1,K1,Up1) 
B2<-bget(Xp2,K2,Up2) 
B3<-bget(Xp3,K3,Up3) 

i=1

plot(Up3[,i], Xp3 %*% B3[,i])
```


```{r}
Ball<-rbind(B1,B2,B3)

```


normalize
```{r}
coef<- t(Ball)
rownames(coef)<- colnames(y)[1:ncol(y)]

barplot(coef)

#coef<-coef[-c(5,6,7),]
barplot(coef[1,])
imp.norm <- diag(1 / apply(coef, 1, sum)) %*% coef
#imp.norm<-scale(coef)
imp.norm<-coef
rownames(imp.norm) <- rownames(coef)

imp.g <- imp.norm[, 1:ncol(genome)]
imp.m <- imp.norm[, (ncol(genome)+1):((ncol(genome))+(ncol(mic.cd)))]
imp.met <- imp.norm[, ((ncol(genome))+(ncol(mic.cd))+1):(ncol(imp.norm))]
barplot(cbind(imp.g, imp.m, imp.met))
```

```{r}
hist(apply(imp.m, 2, sum),main="Sum of coef for each OTU",breaks=100,xlab = "coef of OTU",ylab = "The Number of OTU")
```


## Micro and Met and genome


```{r, fig.width = 7, fig.asp = 1}
imp_all <- cbind(imp.g, imp.m, imp.met)
threshold <- 0
imp_all_filtered <- imp_all[, apply(imp_all, 2, var) > threshold]
dim(imp_all_filtered)


res <- svd(imp_all_filtered)
contr <- res$d / sum(res$d)

rownames(res$u) <- rownames(imp_all_filtered)
rownames(res$v) <- colnames(imp_all_filtered)

ud <- res$u %*% diag(sqrt(res$d))
vd <- res$v %*% diag(sqrt(res$d))

df.u <- data.frame(ud[, 1:2] , Type = "Phenotype", Name = rownames(ud))

type <- rep("Other", nrow(vd))
type[substr(rownames(vd), 1,1) == "X"] <- "Metabolome"
type[substr(rownames(vd), 1,1) == "C"] <- "Genotype"
type[substr(rownames(vd), 1,1) == "B"] <- "Microbiome"
df.v <- data.frame(vd[, 1:2], Type = type, Name = rownames(vd))
df <- data.frame(rbind(df.u, df.v))

colnames(df)[1:2] <- colnames(df.u)[1:2] <- colnames(df.v)[1:2] <- paste0("LF", 1:2)

head(df)

g <- ggplot(df, aes(x = LF1, y = LF2, color = Type)) +

  geom_segment(data = df[1:9, ], 
               aes(x = 0, y = 0, xend = LF1, yend = LF2), 
               arrow = arrow(length = unit(0.2, "inches")), 
               color = "darkgreen", alpha = 0.5) + 

  geom_label_repel(aes(label = Name),
                   box.padding = 0.2, 
                   point.padding = 0.5,
                   segment.color = 'grey50',
                   cex = 2, alpha = 0.9, show.legend = FALSE) +

  geom_point(shape = 21, alpha = 1) +

  scale_color_manual(values = c("#FC4E07", "#E7B800", "#00AFBB", "darkgreen"))

g


```


```{r}
imp_total<-cbind(imp.g,imp.m, imp.met )
saveRDS(imp_total,here("out","imp_BLUP2_c.RDS"))
```



# met


```{r, message =F, include=TRUE,eval=TRUE}
K1 <- tcrossprod(genome) / ncol(genome)
K2 <- tcrossprod(mic.cd) / ncol(mic.cd)

Z <- design.Z(pheno.labels = rownames(genome),geno.names = rownames(genome))

KZ1_2<- list(A= list(K = K1, Z = Z),B= list(K = K2, Z = Z))

```


```{R, message =F, include=TRUE,eval=TRUE}
y <- met.cd

resblup<- foreach(zl = 1) %do% {
  print("fine")
  ZETA <- KZ1_2
  X0 <- NULL
  
  result_matrix <- matrix(nrow = 2*nrow(y), ncol = 0)
  for (i in 1:ncol(y)){
  	print(i)
    resEM3 <- EM3.cpp(y = y[,i],
                        X0 = X0, n.core = 1, 
                        ZETA = ZETA)
    result_matrix <- cbind(result_matrix, resEM3[["u.each"]])
  }
  return(result_matrix)
  }
  
names(resblup)<- "KZ1_2"
```


```{r}
bget <- function(Xp,K,Up) {
  XXt <- K * ncol(Xp)
  B <- t(Xp) %*% ginv(XXt) %*% Up
  return(B)
}

Up<- resblup[["KZ1_2"]]

Up1<-Up[1:nrow(y),]
Up2<-Up[(nrow(y)+1):(nrow(y)*2),]


#Xp1<-genome.cd
Xp1<-genome
Xp2<-mic.cd

B1<-bget(Xp1,K1,Up1) 
B2<-bget(Xp2,K2,Up2) 

```


```{r}
Ball<-rbind(B1,B2)

```


normalize
```{r}
coef<- t(Ball)
rownames(coef)<- colnames(y)[1:ncol(y)]

barplot(coef)

#coef<-coef[-c(5,6,7),]
barplot(coef[1,])
#imp.norm <- diag(1 / apply(coef, 1, sum)) %*% coef
#imp.norm<-scale(coef)
imp.norm<-coef
rownames(imp.norm) <- rownames(coef)

imp.g <- imp.norm[, 1:ncol(genome)]
imp.m <- imp.norm[, (ncol(genome)+1):((ncol(genome))+(ncol(mic.cd)))]
barplot(cbind(imp.g, imp.m))
```




```{r, fig.width = 7, fig.asp = 1}
imp.gm <- cbind(imp.g, imp.m)
#imp.gm <- imp.gm[, colSums(imp.gm) >= 0.002]
res <- svd(imp.gm)
contr <- res$d / sum(res$d)



rownames(res$u) <- rownames(imp.gm)
rownames(res$v) <- colnames(imp.gm)

ud <- res$u %*% diag(sqrt(res$d))
vd <- res$v %*% diag(sqrt(res$d))

df.u <- data.frame(ud[, 1:2] , Type = "Metabolome", Name = rownames(ud))

type <- rep("Other", nrow(vd))
type[substr(rownames(vd), 1,1) == "X"] <- "Metabolome"
type[substr(rownames(vd), 1,1) == "C"] <- "Genotype"
type[substr(rownames(vd), 1,1) == "B"] <- "Microbiome"

df.v <- data.frame(vd[, 1:2], Type = type, Name = rownames(vd))
df <- data.frame(rbind(df.u, df.v))
colnames(df)[1:2] <- colnames(df.u)[1:2] <- colnames(df.v)[1:2] <- paste0("LF", 1:2)
head(df)

colnames(df)[1:2] <- colnames(df.u)[1:2] <- colnames(df.v)[1:2] <- paste0("LF", 1:2)

head(df)

g <- ggplot(df, aes(x = LF1, y = LF2, color = Type))
g <- g + geom_point(shape = 21,alpha=1) 
g <- g + geom_label_repel(aes(label = Name),
                  box.padding   = 0.2, 
                  point.padding = 0.5,
                  segment.color = 'grey50',
                  max.overlaps = 2,
                  cex=2,alpha=0.9,show.legend= FALSE )
g<- g + scale_color_manual(values = c( "#FC4E07","#E7B800","#00AFBB"))
g

```


```{r}
imp_total<-cbind(imp.g,imp.m)
saveRDS(imp_total,here("out","imp_BLUP_met2.RDS"))
```

