---
title: 'Progress Report in UMR MIA Paris-Saclay'
author: "Hayato YOSHIOKA (yoshioka-hayato393@g.ecc.u-tokyo.ac.jp)"
date: "Feb. 1st - Apr. 30th, 2023"
output:
  pdf_document:
    toc: yes
  html_document:
    toc: yes
    df_print: paged
---
```{r setup, include = FALSE}
knitr::opts_chunk$set(eval = TRUE,echo = FALSE,include=FALSE)
```

```{r, message =F}
require(here)
```

\newpage

# Chapter1. Summary

I was invited for a 3 months visit at the UMR MIA Paris-Saclay laboratory from the University of Tokyo to work on my PhD project regarding the Modeling of Plant-Microbe Interactions Using Multi-Omics Data. This project requires the development of new statistical approaches in order to i) study the symbiotic relationship between plants and microorganisms associated to their rhizosphere through genomic data, and i) incorporate this biological prior information into dedicated prediction models in order to improve crop yield prediction. To elaborate the relevant statistical methodology I am co-supervised by Julie AUBERT, who has already developed several approaches for the analysis of plant-microbial communities interactions in the rhizosphere, and Tristan MARY-HUARD who has a strong background in prediction methods applied to quantitative genetics. 
<br>

## 1.1 Background and Motivation
It have been reported Rhizosphere microorganisms protect plants from pests and pathogens and play an important role in plant nutrient absorption and stress tolerance (A. Kristin and H. Miranda, Plant Soil, 2013). Recent advances in the measurement of various omics data have also made it possible to comprehensively acquire and analyze ecological information of plants, such as elemental concentrations (ionome) and metabolite concentrations (metabolome). This new analytical method has revealed interrelationships between the composition of the rhizosphere microbiota and root metabolites (Tenenhaus et al., Psychometrika, 2017). In addition, genomic prediction (GP) (Meuwissen et al., Genetics, 2001), which predicts crop phenotypes based on genome-wide DNA marker information, is widely used. It has also been reported that omics can be used as an intermediate trait for a genomic predction (Christensen et al., 2021). Therefore, studying rhizosphere microorganisms using various omics data (multi-omics data), including genomic information, will lead to a genetic understanding of the symbiotic relationship between plants and rhizosphere microorganisms and to the design of cultivation and breeding methods that utilize the microflora. 

## 1.2.1 What I Did Before the Stay
I have proposed a two-step yield prediction model in which we make use of both the genetic information collected on the soil and the plant variety as intermediate traits (Figure 1). By learning intermediate traits, this model was expected to predict the yield without metabolome data by predicting "Dummy" metabolome dataset. After several attempts, we have demonstrated that our two-step prediction model with 2019 data works well. In other words, we have proven that the predicted "Dummy" met improves prediction accuracy. 

## 1.2.2 What I Did During the Stay
Although we know that our two-step model actually works in the 2019 dataset, we were not sure that the inter-year prediction would also work. To see that, we used 2019 data to build the model and validate it with 2020 data. With the support of MIA-Paris Saclay team, we have shown that our two-step model was actually applicable for the inter-year, which implies the possibility that our model could be used to predict the yield of the next year in real agriculture.

## 1.2.3 Next Goal
A long stay is planned starting in August 2023. Based on the results of this short stay, we will work on further model development. In particular, I am aiming to develop models with latent variables, as they have the potential to improve this prediction.

```{r,include=TRUE, fig.align = 'center', out.width = "100%", fig.cap='\\label{fig:figs} Two-step Model; First, the met is predicted from the upstream and replaced with the predicted values. Secondly, the yield is predicted from the G, Micro, and Predicted "Dummy" Met'}

knitr::include_graphics(here("image", "inrae1.png"))

```

<br>

## 1.3 Materials and Methods
For the analysis, the following soybean multi-omics data obtained from a field in Tottori, Japan were used. We prepared whole genome sequences of 198 soybean lines (soybean genome: G), elemental concentrations in fully expanded leaves at harvest (leaf ionome: LI) and metabolite concentrations (leaf metabolome: LM), root metabolite concentrations in rhizosphere microflora (meta-metabolome: Met), 16S sequence data of rhizosphere microflora (microbial metagenome: MG), and soybean phenotype (P) data. The complete dataset is available for the 2019 and 2020 data; for 2018, we do not have the meta-metabolome, but we have other omics and trait data (Figure 2,3). For each omics, the inner product of each variable was calculated to obtain a linear kernel. The calculated kernel is generally called Genomic Relationship Matrix (GRM) for genome. In here, the kernel for the Meta-Metabolome is called MRM, and the kernel for the Microbiome is called MicroRM. Genomic best linear unbiased prediction (gBLUP) is a method that utilizes genomic relationships to estimate the genetic merit of an individual. For this purpose, a genomic relationship matrix is used, estimated from DNA marker information. The matrix defines the covariance between individuals based on observed similarity at the genomic level, rather than on expected similarity based on pedigree, so that more accurate predictions of merit can be made. In this study, the model was limited to a classic BLUP in order to test the accuracy of prediction by linear model. The equation of BLUP is shown below.

$$
y = X\beta + Zu+\epsilon
$$

Where $y$ is a vector of phenotype, $X$ is a design matrix relating the fixed effects to each animal, $\beta$ is a vector of fixed Effects, $Z$ is a design matrix allocating records to genetic values, $u$ is a vector of additive genetic effects for an individual, and $\epsilon$ is a vector of random normal deviates with variance $\sigma_{e}^{2}$. Especially, when you perform GBLUP,


$$ 
Var\left( u \right) = G\sigma_{g}^{2}
$$

where $G$ is the genomic relationship matrixc (GRM), and $\sigma_{g}^{2}$ is the genetic variance for this model. Note that we also perform BLUP with Meta-Metabolome Relationship Matrix (MRM) and Microbiome Relationship Matrix (MicroRM). Since their variance can be calculated just like Genomic data, we applied multi-kenerl model to the multi-omics prediction. I used "EM3.cpp" function from a R package "RAINBOWR" (Hamazaki & Iwata, 2020), which solves the following multi-kernel linear mixed effects model. 

$$
y = X\beta + \Sigma_{l=1}^{L}Z_{l}u_{l}+\epsilon \hspace{20pt} where \hspace{5pt} Var[y] = \Sigma_{l=1}^{L}Z{l}K_{l} Z_{l}'\sigma_{l}^2 + I\sigma_{e}^{2}. 
$$

To see if the two-step prediction model can be applied to the coming year data, we used data from 2019 and 2020 to build and validate it.


```{r,include=TRUE, fig.align = 'center', out.width = "100%", fig.cap="\\label{fig:figs}Materials: Phenotypic data and 5 omics datasets. We prepared whole genome sequences of 198 soybean lines (soybean genome: G), elemental concentrations in fully expanded leaves at harvest (leaf ionome: LI) and metabolite concentrations (leaf metabolome: LM), root metabolite concentrations in rhizosphere microflora (meta-metabolome: Met), 16S sequence data of rhizosphere microflora (microbial metagenome: MG), and soybean phenotype (P) data"}

knitr::include_graphics(here("image", "inrae0.png"))

```

```{r,include=TRUE, fig.align = 'center', out.width = "80%", fig.cap="\\label{Table:1}Available Datasets.The complete dataset is available for the 2019 and 2020 data; for 2018, we do not have the meta-metabolome, but we have other omics and trait data"}

knitr::include_graphics(here("image", "inrae4.png"))
```


\newpage

# Chapter 2. Met Prediciton by BLUP in 2019 and 2020

For the prediction of Met, We set the following models\

2.1 : G -> Met (2019 to 2020)\
2.1 : G + Micro -> Met (2019 to 2020)\
2.2 : G + Micro -> Met (2020 to 2020)


For the Met prediction, we have built the following BLUP model.\
$$
y = X\beta + \Sigma_{l=1}^{L}Z_{l}u_{l}+\epsilon
$$

where $y$ is a vector of phenotype, which is a concentration value for metabolites, $X$ is a design matrix relating the fixed effects to each conditions or year, $\beta$ is a vector of fixed Effects (we did not set any fixed effects for this study because we wanted to predict from 2019 to 2020 without mixing them), $Z$ is a design matrix allocating records to genetic or microbiomic values, $u$ is a vector of additive genetic or microbiomic effects for an individual, and $\epsilon$ is a vector of random normal deviates with variance $\sigma_{e}^{2}$.

## 2.1 Inter-year Met Prediction 
We have built a BLUP model to predict 2020 met data from 2019 data. By masking the 2020 phenotypic value, the model was learned based on the only 2019 dataset. BLUP was performed for each metabolite each time. It means we performed BLUP 264 times (264 is the number of metabolites). $N$ is the number of the plant lines (N=148), Met is a matrix of Metabolome data (2N × 264), The size of GRM is N × N and the size of MicroRM is 2N × 2N. However, to use the same design matrix for all models, i set 2N × 2N GRM, it has 4 GRM matrice in it. (Figure 4)

```{r,include=TRUE, fig.align = 'center', out.width = "100%", fig.cap="\\label{fig:figs} The prediction of Dummy Met. $N$ is the number of the plant lines (N=148), Met is a matrix of Metabolome data (2N × 264), The size of GRM is N × N and the size of MicroRM is 2N × 2N."}

knitr::include_graphics(here("image", "inrae2.png"))
```


## 2.2 One-year Met Prediction in 2020
For One-year Met prediction, we performed the leave-one-out cross validation to predict each metabolite on each line. It means we performed BLUP 39072 (= 148 × 264) times. $N$ is the number of the plant lines (N=148), Met is a matrix of Metabolome data (N × 264), The size of GRM is N × N and the size of MicroRM is N × N. (Figure 5)

```{r,include=TRUE, fig.align = 'center', out.width = "100%", fig.cap="\\label{fig:figs}The prediction of Dummy Met. $N$ is the number of the plant lines (N=148), Met is a matrix of Metabolome data (N × 264), The size of GRM and MicroRM is N × N."}

knitr::include_graphics(here("image", "inrae3.png"))

```

## 2.3 Results of Met Prediction

Here, the correlation coefficients between the predicted and true Met concentration values are shown in the figure.\
G -> Met (2019 to 2020)\
G + Micro -> Met (2019 to 2020)\
G + Micro -> Met (2020 to 2020)

```{r child = '50_INRAErepo1.Rmd'}
```

When I put both GRM and MicroRM into the model, we could get better results rather than putting only GRM, which means we could put environmental effect to the model by using MicroRM. Compared to inter-year model, one-year model showed higher accuracy, but it has some outlier in Met prediction. On the other hand, the inter-year predction didn't have outlier. (Figure 6) The outlier might effect on the prediction of the yield in the down streeam of the two step model.

\newpage

# Cahpter 3. Yield Prediction in 2020

## 3.1 Model Build for 2020 Yield Prediction

For the inter-year prediction problem, we set the following prediction rules (PR).\
- the PR based on Genetic information only, ie the classical GBLUP\
- the PR based on Genetic and Micro\
- the PR based on Genetic, Micro and Meta\
- the PR based on Genetic, Micro and predicted Meta based on Genetic and Micro

For the yield prediction, we have built the following BLUP model.\
$$
y = X\beta + \Sigma_{l=1}^{L}Z_{l}u_{l}+\epsilon
$$

where $y$ is a vector of phenotype, which is the yield, $X$ is a design matrix relating the fixed effects to each conditions or year, $\beta$ is a vector of fixed Effects (we did not set any fixed effects for this study because we wanted to predict from 2019 to 2020 without mixing them), $Z$ is a design matrix allocating records to genetic, microbiomic, or metabolomic values, $u$ is a vector of additive genetic, microbiomic, or metabolomic effects for an individual, and $\epsilon$ is a vector of random normal deviates with variance $\sigma_{e}^{2}$. Note that I got inear kernel for predicted Metabolome matrix and True Met matrix as well as GRM and MicroRM.

## 3.2 Results of Yield Prediction

The results of the yield prediction in 2020 is shown below. 

```{r child = '51_INRAErepo2.Rmd'}
```

PredMet_f19 represents Metabolome data predicted by 2019 dataset. Now, G + Micro + PredMet_f19 was higher than G + Micro, which implies met predicted by the inter-year model contributes to the yield prediction. On the other hand, the one-year model didnt work well. Notably, The accuracy of iter-year model was better than inter-year, and the reason is mentioned in discussion.

\newpage

# Chapter 4. Discussion
## Inter-year and One-year comparison 
The accuracy of one-year model was worse than inter-year. This may be due to leave-one-out prediction method had significant effects on the model accuracy. In here, we did not perform leave-one-out for the inter year. Thus we need to validate both in the leave-one-out method when we compared them. (This is actually tested just after the short stay. The result shows that one-year accuracy was better than inter-year. As for most traits, Met predicted by the One-year model improved the accuracy, but one by the Inter-year model didn't. However, if we consider the only inter-year model, we do NOT need to perform Leave-One-Out. Thus, It was meaningful that met predicted by inter-year model significantly improved the model in this case. 

## Genetic and Microbiomic variance
We need to be careful that the predicted met could be almost copied from the previous year's one. To avoid that, we have put the microbiome information (which are different depending on the year), but it is possible that met can be mainly controlled by Genomic information according to the result of Chapter2.

## outlier
One-year model showed higher accuracy, but it has some outlier in Met prediction (Figure). I previously tested the Random Forest model for Met predictions for the one-year model and found fewer outlier (Figure 9). Thus, It is possible BLUP model was not flexible enough to explain the complex relationship between plants, microbiome, and metabolites; we need to consider which model to use for the one-year model to reduce them. Some metabolites prediction failed to BLUP in G + Micro -> Met model (2020->2020). Because BLUP showed high accuracy for some metabolites, metabolites showing abnormal values may have different concentration trends. The result of outlier is shown below (Figure 10).

```{r,include=TRUE, fig.align = 'center', out.width = "100%", fig.cap="\\label{fig:figs} The predicted Dummy Met by the one-year model in 2019 (2019->2019) "}

knitr::include_graphics(here("image", "inrae5.png"))
```

## Future Goal
First, the model trained in 2019 and 2020 will be used to predict the 2018 yield, and other combinations will be tried to validate the inter-year models further. Secondly, we will test possible statistical models such as the latent variable model and machine learning. The result shows that BLUP may be unable to explain the complex relationships among omics. Thus, we need to consider another statistical approach for the prediction. For example, we could apply a multivariate model for the prediction to include the relationship between each metabolite, which may improve the accuracy. Finally, we will develop a new latent variable model. Dimensionality reduction is crucial in omics research, and several latent variable models have been developed in the field of biostatistics. Based on PCA and CCA, I want to develop a latent variable method that incorporates non-linear information so that the model can explain the complex interactions between omics information. I would also like to evaluate the performance of Autoencoder.


```{R fig ,include=TRUE,fig.dim = c(8, 5),fig.cap="\\label{fig:figs} outlier in Met prediction from 2020 to 2020. In some metabloites, the fitting clearly failed. rel line represent y=x"}
predmet20 <- read.csv(here("2020", "2020_predmet_blup.csv"), row.names = 1)
met20.cd <- read.csv(here("2020", "met20d.csv"), row.names = 1)
cor<- cor(predmet20,met20.cd)
pc<- diag(cor)
pcsel<- (pc < -0.999)
#pcsel<- (pc > 0.5)
t<- table(pcsel)

par(mfrow = c(2, 4))
for (i in 1:t["TRUE"]){

plot(scale(predmet20[,pcsel][,i]),met20.cd[,pcsel][,i],xlab = "PredMet20from20",ylab="Met20")
abline(a=0,b=1,col=2)
#met20.cd[met20.cd[,i] > 10,i]
  }
```




# Reference
Christensen, O. F., Börner, V., Varona, L., & Legarra, A. (2021). Genetic evaluation including intermediate omics features. Genetics, 219(2). https://doi.org/10.1093/genetics/iyab130\

Hamazaki, K., & Iwata, H. (2020). Rainbow: Haplotype-based genome-wide association study using a novel SNP-set method. PLoS Computational Biology, 16(2), 1–17. https://doi.org/10.1371/journal.pcbi.1007663\

Kristin, A., & Miranda, H. (2013). The root microbiota-a fingerprint in the soil? In Plant and Soil (Vol. 370, Issues 1–2, pp. 671–686). https://doi.org/10.1007/s11104-013-1647-7\

Meuwissen, T. H. E., Hayes, B. J., & Goddard, M. E. (2001). Prediction of Total Genetic Value Using Genome-Wide Dense Marker Maps. https://academic.oup.com/genetics/article/157/4/1819/6048353\

Tenenhaus, M., Tenenhaus, A., & Groenen, P. J. F. (2017). Regularized Generalized Canonical Correlation Analysis: A Framework for Sequential Multiblock Component Methods. Psychometrika, 82(3), 737–777. https://doi.org/10.1007/s11336-017-9573-x\


